src/iop/expfusion.c

```diff
@@ -0,0 +1,1268 @@
+    else
+      exp = 0.f;
+  }
+
+  return exp;
+}
+
+static void _buil_weight_map(const float *const img_src, const size_t wd, const size_t ht, const int ch,
+                             float *const img_map, const dt_iop_grey_projectors_t grey_projector, const dt_iop_weight_modes_t weight_mode,
+                             const float exposure_optimum, const float exposure_width,
+                             const float exposure_left_cutoff, const float exposure_right_cutoff,
+                             const dt_iop_order_iccprofile_info_t *const work_profile)
+{
+#ifdef _OPENMP
+#pragma omp parallel for default(none) \
+  dt_omp_firstprivate(img_src, img_map, ht, wd, ch, grey_projector, work_profile, weight_mode, \
+  exposure_optimum, exposure_width) \
+  schedule(static) \
+  collapse(2)
+#endif
+  for(size_t y = 0; y < ht; y++)
+  {
+    for(size_t x = 0; x < wd; x++)
+    {
+      const float *const rgb = img_src + (y * wd * ch) + x * ch;
+
+      float E = 1.f;
+      if(grey_projector != DT_PROJECTOR_NONE)
+      {
+        const float lum = _grey_projector(rgb, grey_projector, work_profile);
+        E = _well_exposedness(lum, weight_mode, exposure_optimum, exposure_width, exposure_left_cutoff, exposure_right_cutoff);
+      }
+      else
+      {
+        E = _well_exposedness(rgb[0], weight_mode, exposure_optimum, exposure_width, exposure_left_cutoff, exposure_right_cutoff) *
+              _well_exposedness(rgb[1], weight_mode, exposure_optimum, exposure_width, exposure_left_cutoff, exposure_right_cutoff) *
+              _well_exposedness(rgb[2], weight_mode, exposure_optimum, exposure_width, exposure_left_cutoff, exposure_right_cutoff);
+      }
+
+      img_map[y * wd + x] = E;
+    }
+  }
+}
+
+static float exposure_increment(const float stops, const int e)
+{
+  const float white = exp2f(-stops * (float)e);
+  const float scale = 1.0f / white;
+  return scale;
+}
+
+static void _image_rgb_to_grey(const float *const img_src, const size_t wd, const size_t ht, const int ch, float *const img_dest,
+                                const dt_iop_grey_projectors_t grey_projector, const dt_iop_order_iccprofile_info_t *const work_profile)
+{
+  const size_t size = wd * ht * ch;
+
+#ifdef _OPENMP
+#pragma omp parallel for default(none) \
+  dt_omp_firstprivate(img_src, img_dest, size, ch, grey_projector, work_profile) \
+  schedule(static)
+#endif
+  for(size_t i = 0; i < size; i += ch)
+    img_dest[i] = img_dest[i + 1] = img_dest[i + 2] = _grey_projector(img_src + i, grey_projector, work_profile);
+}
+
+static void _exposure_fusion(const float *const img_src, const size_t wd, const size_t ht, const int ch,
+                             const int num_exposures, float *const img_dest, struct dt_iop_module_t *self,
+                             const dt_iop_order_iccprofile_info_t *const work_profile,
+                             dt_iop_fusion_data_t *const d)
+{
+  const int num_levels = floor(log2(MIN(wd, ht)));
```

Comment:
Increasing num_levels definitely has significant benefits as far as haloing, although ideally there should be a testing goal to be approximately on par with or better than enfuse when num_levels is 8 - otherwise that's a strong indication that num_levels is masking an underlying flaw in the algorithm.

Also, I looked briefly at the tiling callbacks in the old module, ideal values of num_levels could present a very nasty challenge on memory-constrained systems by requiring excessive amounts of tile overlap.  I'm wondering if that is why num_levels was capped at 8 in the original implementation - which would limit tile overlap to 256 pixels.